### PR TITLE
Zoom In/Out keyboard shortcut

### PR DESCRIPTION
**Pull request by Quasar985**
### Prerequisites

- [x] Reviewed the [checklist](CHECKLIST.md)

- [x] Reviewed feedback from the "Sonar Cloud" bot. Note that you have to wait
    for the "CI / Unit Tests") to complete first. Failed Unit tests can be 
    debugged by adding the label "verbose logging" to the GitHub PR.

### Description of the Change

Added keyboard shortcuts and toolbar buttons for zooming in/out the graph view. Shortcuts are Ctrl+Alt+Up and Ctrl+Alt+Down, respectively.

### Alternate Designs

Buttons could be positioned in different areas of the UI. Their current position is close to buttons of similar funciton, "Reset View" and "Zoom To Selection" on the toolbar.

### Why Should This Be In Core?

Allows users without a scroll wheel to conveniently zoom in and out of the graph view.

### Benefits

Greater accessibility to users of different hardware configurations.

### Possible Drawbacks

Toolbar UI may eventually become cluttered if more buttons are added.

### Verification Process

Combination of visual testing in graph view and unit tests written to verify camera position.

### Applicable Issues

https://github.com/constellation-app/constellation/issues/1955

### CODE REVIEW

File: CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/plugins/zoom/ZoomOutPlugin.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+ *
+ * @author Quasar985
+ */
+@ServiceProvider(service = Plugin.class)
+@NbBundle.Messages("ZoomOutPlugin=Zoom Out")
+@PluginInfo(minLogInterval = 5000, pluginType = PluginType.VIEW, tags = {PluginTags.VIEW})
+public final class ZoomOutPlugin extends SimpleEditPlugin {
+
+    final Vector3f ZOOM_DIRECTION = new Vector3f(0, 0, -1);
+    final int ZOOM_AMOUNT = 10;
```
Comment (capricornunicorn123) [inline CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/plugins/zoom/ZoomOutPlugin.java:48]: These two variables can be declared static

File: CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/plugins/zoom/ZoomInPlugin.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+ *
+ * @author Quasar985
+ */
+@ServiceProvider(service = Plugin.class)
+@NbBundle.Messages("ZoomInPlugin=Zoom In")
+@PluginInfo(minLogInterval = 5000, pluginType = PluginType.VIEW, tags = {PluginTags.VIEW})
+public final class ZoomInPlugin extends SimpleEditPlugin {
+
+    final Vector3f ZOOM_DIRECTION = new Vector3f(0, 0, 1);
+    final int ZOOM_AMOUNT = 10;
```
Comment (capricornunicorn123) [inline CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/plugins/zoom/ZoomInPlugin.java:48]: These two variables can be declared static

File: CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/plugins/zoom/ZoomOutPlugin.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    public void edit(final GraphWriteMethods graph, final PluginInteraction interaction, final PluginParameters parameters) throws InterruptedException, PluginException {
+        final Camera oldCamera = VisualGraphUtilities.getCamera(graph);
+        final Camera camera = new Camera(oldCamera);
+
+        CameraUtilities.zoom(camera, ZOOM_AMOUNT, ZOOM_DIRECTION, 0F);
+
+        final Graph activeGraph = GraphManager.getDefault().getActiveGraph();
+        if (activeGraph != null && activeGraph.getId().equals(graph.getId())) {
+            // Only do the camera animation if the edited graph is currently active
+            Animation.startAnimation(new PanAnimation("Zoom to Selection", oldCamera, camera, true));
```
Comment (capricornunicorn123) [inline CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/plugins/zoom/ZoomOutPlugin.java:60]: Running this as an animation looks nice when users zoom with a gap between zoom requests. However, when zooming quicky. (i.e ctrl + alt + hold up arrow) the experiences a little jarring with pan animations stopping and starting. I wonder if simply writing to the graph and not doing an animation would be nicer? Single requests to zoom would be more jarring but might be a lesser evil. (Same goes for ZoonInPlugin)

File: CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/plugins/zoom/ZoomOutPlugin.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+    final Vector3f ZOOM_DIRECTION = new Vector3f(0, 0, -1);
+    final int ZOOM_AMOUNT = 10;
+
+    @Override
+    public void edit(final GraphWriteMethods graph, final PluginInteraction interaction, final PluginParameters parameters) throws InterruptedException, PluginException {
+        final Camera oldCamera = VisualGraphUtilities.getCamera(graph);
+        final Camera camera = new Camera(oldCamera);
+
+        CameraUtilities.zoom(camera, ZOOM_AMOUNT, ZOOM_DIRECTION, 0F);
```
Comment (capricornunicorn123) [inline CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/plugins/zoom/ZoomOutPlugin.java:55]: Setting the node distance to 0F will mean that zooming through vast empty space to distant nodes will be slow. Might be worth trying to get node distances to help with this.

**Review by capricornunicorn123 — COMMENTED** [review]
I wonder if these plugins could be merged to a simple zoom plugin and parameters added to indicate direction, distance etc? This could then be integrated into the `DefaultInteractionEventHandeler` and become a reusable zooming functionality.

**Review by OrionsGuardian — CHANGES_REQUESTED** [review]
We should probably be re-using the existing code for zooming in and out via the mouse wheel, and develop a modified version to handle the keyboard shortcuts.
`DefaultInteractionEventHandler.mouseWheelMoved(event)`

This should also address some of the concerns raised by @capricornunicorn123

File: CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/plugins/zoom/ZoomInPlugin.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+ *
+ * @author Quasar985
+ */
+@ServiceProvider(service = Plugin.class)
+@NbBundle.Messages("ZoomInPlugin=Zoom In")
+@PluginInfo(minLogInterval = 5000, pluginType = PluginType.VIEW, tags = {PluginTags.VIEW})
+public final class ZoomInPlugin extends SimpleEditPlugin {
+
+    static private final Vector3f ZOOM_DIRECTION = new Vector3f(0, 0, 1);
+    static private final int ZOOM_MAGNITUDE = 2;
```
Comment (Delphinus8821) [inline CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/plugins/zoom/ZoomInPlugin.java:41]: private static

File: CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/plugins/zoom/ZoomOutPlugin.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+ *
+ * @author Quasar985
+ */
+@ServiceProvider(service = Plugin.class)
+@NbBundle.Messages("ZoomOutPlugin=Zoom Out")
+@PluginInfo(minLogInterval = 5000, pluginType = PluginType.VIEW, tags = {PluginTags.VIEW})
+public final class ZoomOutPlugin extends SimpleEditPlugin {
+
+    static private final Vector3f ZOOM_DIRECTION = new Vector3f(0, 0, -1);
+    static private final int ZOOM_MAGNITUDE = 2;
```
Comment (Delphinus8821) [inline CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/plugins/zoom/ZoomOutPlugin.java:41]: private static

File: CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/plugins/zoom/ZoomUtilities.java [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+     * Function for zooming a camera's view by a given amount in a given direction.
+     *
+     * @param graph the graph that holds the camera
+     * @param zoomMagnitude the amount of units to zoom by
+     * @param zoomDirection the direction to zoom to as a 3d vector
+     * @param distanceToClosestNode the distance to the closest node, used to influence zoom amount
+     * @throws InterruptedException
+     * @throws PluginException
+     */
+    static public void zoom(final GraphWriteMethods graph, final int zoomMagnitude, final Vector3f zoomDirection, float distanceToClosestNode) throws InterruptedException, PluginException {
```
Comment (Delphinus8821) [inline CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/plugins/zoom/ZoomUtilities.java:44]: public static and last parameter final

File: CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/plugins/zoom/ZoomUtilities.java [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+     * Function for zooming a camera's view by a given amount in a given direction. Variation of zoom that calculates
+     * the distanceToClosestNode itself
+     *
+     * @param graph the graph that holds the camera
+     * @param zoomMagnitude the amount of units to zoom by
+     * @param zoomDirection the direction to zoom to as a 3d vector
+     * @throws InterruptedException
+     * @throws PluginException
+     */
+    static public void zoom(final GraphWriteMethods graph, final int zoomMagnitude, final Vector3f zoomDirection) throws InterruptedException, PluginException {
```
Comment (Delphinus8821) [inline CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/plugins/zoom/ZoomUtilities.java:66]: public static

File: CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/plugins/zoom/ZoomUtilities.java [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+
+        zoom(graph, zoomMagnitude, zoomDirection, distanceToClosestNode);
+    }
+
+    /**
+     * Function for finding the closest node to the camera in a given graph
+     *
+     * @param graph the graph that holds the camera and nodes
+     */
+    static public Vector3f closestNodeCameraCoordinates(GraphReadMethods graph) {
```
Comment (Delphinus8821) [inline CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/plugins/zoom/ZoomUtilities.java:81]: public static and parameter final

File: CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/visual/InteractiveGLVisualProcessor.java [hunk 10]
```diff
@@ -332,7 +335,11 @@ public NodeCameraDistance(Vector3f nodeLocation, final float horizontalScale, fi
             this.distanceFromCamera = getDistanceFromCamera(nodeLocation, horizontalScale, verticalScale);
         }
 
-        static NodeCameraDistance getClosestNode(NodeCameraDistance ncd1, NodeCameraDistance ncd2) {
+        public Vector3f GetNodeLocation() {
+            return nodeLocation;
+        }
+
+        static public NodeCameraDistance getClosestNode(NodeCameraDistance ncd1, NodeCameraDistance ncd2) {
```
Comment (Delphinus8821) [inline CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/visual/InteractiveGLVisualProcessor.java:342]: public static and parameters final

**Review by Delphinus8821 — CHANGES_REQUESTED** [review]

**Review by Delphinus8821 — APPROVED** [review]

File: CoreInteractiveGraph/test/unit/src/au/gov/asd/tac/constellation/graph/interaction/plugins/zoom/ZoomInPluginNGTest.java [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+import au.gov.asd.tac.constellation.graph.StoreGraph;
+import au.gov.asd.tac.constellation.graph.schema.Schema;
+import au.gov.asd.tac.constellation.graph.schema.SchemaFactoryUtilities;
+import au.gov.asd.tac.constellation.graph.schema.analytic.AnalyticSchemaFactory;
+import au.gov.asd.tac.constellation.graph.schema.visual.concept.VisualConcept;
+import au.gov.asd.tac.constellation.plugins.PluginInteraction;
+import au.gov.asd.tac.constellation.plugins.parameters.PluginParameters;
+import au.gov.asd.tac.constellation.utilities.camera.Camera;
+import au.gov.asd.tac.constellation.utilities.graphics.Vector3f;
+import static org.testng.Assert.*;
```
Comment (antares1470) [inline CoreInteractiveGraph/test/unit/src/au/gov/asd/tac/constellation/graph/interaction/plugins/zoom/ZoomInPluginNGTest.java:27]: can you replace this import with the specifically used classes?

File: CoreInteractiveGraph/test/unit/src/au/gov/asd/tac/constellation/graph/interaction/plugins/zoom/ZoomOutPluginNGTest.java [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+import au.gov.asd.tac.constellation.graph.StoreGraph;
+import au.gov.asd.tac.constellation.graph.schema.Schema;
+import au.gov.asd.tac.constellation.graph.schema.SchemaFactoryUtilities;
+import au.gov.asd.tac.constellation.graph.schema.analytic.AnalyticSchemaFactory;
+import au.gov.asd.tac.constellation.graph.schema.visual.concept.VisualConcept;
+import au.gov.asd.tac.constellation.plugins.PluginInteraction;
+import au.gov.asd.tac.constellation.plugins.parameters.PluginParameters;
+import au.gov.asd.tac.constellation.utilities.camera.Camera;
+import au.gov.asd.tac.constellation.utilities.graphics.Vector3f;
+import static org.testng.Assert.*;
```
Comment (antares1470) [inline CoreInteractiveGraph/test/unit/src/au/gov/asd/tac/constellation/graph/interaction/plugins/zoom/ZoomOutPluginNGTest.java:27]: same here

File: CoreInteractiveGraph/test/unit/src/au/gov/asd/tac/constellation/graph/interaction/plugins/zoom/ZoomUtilitiesNGTest.java [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+
+import au.gov.asd.tac.constellation.graph.GraphWriteMethods;
+import au.gov.asd.tac.constellation.graph.StoreGraph;
+import au.gov.asd.tac.constellation.graph.schema.Schema;
+import au.gov.asd.tac.constellation.graph.schema.SchemaFactoryUtilities;
+import au.gov.asd.tac.constellation.graph.schema.analytic.AnalyticSchemaFactory;
+import au.gov.asd.tac.constellation.graph.schema.visual.concept.VisualConcept;
+import au.gov.asd.tac.constellation.utilities.camera.Camera;
+import au.gov.asd.tac.constellation.utilities.graphics.Vector3f;
+import static org.testng.Assert.*;
```
Comment (antares1470) [inline CoreInteractiveGraph/test/unit/src/au/gov/asd/tac/constellation/graph/interaction/plugins/zoom/ZoomUtilitiesNGTest.java:26]: same here

**Review by antares1470 — COMMENTED** [review]
1) If you would add an entry to `whatsnew.txt` for this, that would be great
2) I see there is a use of `CameraUtiltities.zoom` in the `DefaultInteractionEventHandler` (which is where it handles mouse wheel scrolling). Wondering if that should be altered to use the `ZoomUtilities` class you created

**Comment by Quasar985** [comment]
> 1. If you would add an entry to `whatsnew.txt` for this, that would be great
> 2. I see there is a use of `CameraUtiltities.zoom` in the `DefaultInteractionEventHandler` (which is where it handles mouse wheel scrolling). Wondering if that should be altered to use the `ZoomUtilities` class you created

The `ZoomUtilities` class was more to streamline the shared functionality of `ZoomInPlugin` and `ZoomOutPlugin`, and gather the parameters needed to call `CameraUtiltities.zoom` with what info plugins are able to access. I think swapping out `CameraUtilities` for `ZoomUtilities`, would make `DefaultInteractionEventHandler` slightly worse in terms of code readability and how the graph view is updated.

I'm thinking the solution might be to rename `ZoomUtilities` to better reflect it's intended purpose as a base class for the zoom plugins. Let me know your thoughts.

**Comment by antares1470** [comment]
> The `ZoomUtilities` class was more to streamline the shared functionality of `ZoomInPlugin` and `ZoomOutPlugin`, and gather the parameters needed to call `CameraUtiltities.zoom` with what info plugins are able to access. I think swapping out `CameraUtilities` for `ZoomUtilities` would make `DefaultInteractionEventHandler` slightly worse in terms of code readability and how the graph view is updated.
> 
> I'm thinking the solution might be to rename `ZoomUtilities` to better reflect it's intended purpose as a base class for the zoom plugins. Let me know your thoughts.

I suppose the thinking in my head was that you would expect that the mouse scrolling and the zoom plugins would all be tapping into the same functionality. And they are...but possibly not as much as you would expect. That being said, having looked at the functions, I can see why it might be difficult to consolidate further. Renaming probably also won't help as its likely to be difficult to come up with a better name (even something like "ZoomPluginUtilities" doesn't really remove the confusion).

So based on that, I guess you can ignore that 2nd point (though maybe adding a description to the ZoomUtilities class could help to outline its purpose).

**Review by antares1470 — APPROVED** [review]

**Review by Delphinus8821 — APPROVED** [review]

**Pull request merged**